internal/exec/stages/files/units.go

```diff
@@ -19,35 +19,73 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/coreos/ignition/v2/config/shared/errors"
 	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
 	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/exec/util"
+	"github.com/coreos/ignition/v2/internal/systemd"
 )
 
+// Preset holds the information about
+// a given systemd unit.
+type Preset struct {
+	unit            string
+	enabled         bool
+	instantiateable bool
+	instances       []string
+}
+
+// warnOnOldSystemdVersion checks the version of Systemd
+// in a given system and prints a warning if older than 240.
+func (s *stage) warnOnOldSystemdVersion() error {
+	systemdVersion, err := systemd.GetSystemdVersion()
+	if err != nil {
+		return err
+	}
+	if systemdVersion < 240 {
+		if err := s.Logger.Warning("The version of systemd (%q) is less than 240. Enabling/disabling instantiated units may not work. See https://github.com/coreos/ignition/issues/586 for more information.", systemdVersion); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // createUnits creates the units listed under systemd.units.
 func (s *stage) createUnits(config types.Config) error {
-	enabledOneUnit := false
+	presets := make(map[string]*Preset)
 	for _, unit := range config.Systemd.Units {
 		if err := s.writeSystemdUnit(unit, false); err != nil {
 			return err
 		}
 		if unit.Enabled != nil {
 			if *unit.Enabled {
-				if err := s.Logger.LogOp(
-					func() error { return s.EnableUnit(unit) },
-					"enabling unit %q", unit.Name,
-				); err != nil {
-					return err
+				if strings.Contains(unit.Name, "@") {
+					unitName, instance, err := parseInstanceUnit(unit)
+					if err != nil {
+						return err
+					}
+					if _, ok := presets[unitName]; ok {
+						presets[unitName].instances = append(presets[unitName].instances, instance)
```

Comment:
Is it possible for the same unit name to have multiple instances wanting different enable states? If so this logic would fall apart.

Reply:
I have performed some testing around this particular case, and it looks like `ignition-validate` tool itself will fail to validate the schema:
e.g.:
```
config:
variant: fcos
version: 1.0.0
passwd:
  users:
    - name: core
      ssh_authorized_keys:
        - ssh-rsa AAAA..
systemd:
  units:
    - name: echo@.service
      contents: |
        [Unit]
        Description=f
        [Service]
        Type=oneshot
        ExecStart=/bin/echo %i
        RemainAfterExit=yes
        [Install]
        WantedBy=multi-user.target
    - name: echo@foo.service
      enabled: true
    - name: require-echo-foo.service
      enabled: true
    - name: echo@foo.service
      enabled: false
$ ./ignition-validate simple.ign 
error at $.systemd.units.3, line 40 col 8: duplicate entry defined

```
So practically, IMO that case isn't possible.

Reply:
`unitName` is the first result of `parseInstanceUnit` which is built via `unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]`. So couldn't `echo@foo.service` & `echo@bar.service` both result in the `unitName` being `echo@.service`?

Reply:
Yes, you're right and that's the expected behavior. We don't want to create a separate unit name for each of the instantiated units sharing the same `unitName`.
For e.g
1. valid case
```
 enable echo@.service foo bar
```
2. Invalid case:
You can't have two enable states present in a given ignition file. As I stated in my previous comment, the `ignition-validate` tool will throw an error with `duplicate entry`
```
enable echo@.service foo bar
disable echo@foo.service
disable echo@bar.service
```

Reply:
Okay, so just to make sure that I'm understanding you're saying that this would not be a valid config?
```
config:
variant: fcos
version: 1.0.0
passwd:
  users:
    - name: core
      ssh_authorized_keys:
        - ssh-rsa AAAA..
systemd:
  units:
    - name: echo@.service
      contents: |
        [Unit]
        Description=f
        [Service]
        Type=oneshot
        ExecStart=/bin/echo %i
        RemainAfterExit=yes
        [Install]
        WantedBy=multi-user.target
    - name: echo@foo.service
      enabled: true
    - name: require-echo-foo.service
      enabled: true
    - name: echo@bar.service
      enabled: false
```

Reply:
IIUC correctly this is a valid config

```
    - name: echo@foo.service
      enabled: true
    - name: echo@bar.service
      enabled: false
```

while this is not:

```

    - name: echo@foo.service
      enabled: true
    - name: echo@foo.service
      enabled: false
```

Reply:
> IIUC correctly this is a valid config
> 
> ```
>     - name: echo@foo.service
>       enabled: true
>     - name: echo@bar.service
>       enabled: false
> ```

Ok, so the current code will need to change then to consider the instance enablement state. As it is right now it would just perform the enablement state of the first instance for everything under a given `unitName`.

Reply:
right. should we just update the key in the map to be `unitname` + `unit.Enabled`, which would make it so we could have `presets["echo@.service-enabled"]` `presets["echo@.service-disabled"]`

Reply:
Fixed.